File: glp_runtime/lib/link/seam/link_id.dart
import 'link_address.dart';
import 'link_scheme.dart';

/// The per-establishment uniqueness nonce. Mirrors the GLP
/// `Nonce ::= Integer ; String` component
/// (contracts/link-primitives.md §1, line 56). The integer and string forms are
/// kept distinct (an integer `5` is NOT the string `"5"`) so the value
/// equality that backs idempotency-at-identity (FR-007) matches the GLP term
/// exactly rather than a lossy canonical text.
class LinkNonce {
  /// True when this is the `Integer` form; false for the `String` form.
  final bool isInteger;

  /// The integer value, valid only when [isInteger].
  final int intValue;

  /// The string value, valid only when [isInteger] is false.
  final String? stringValue;

  const LinkNonce._(this.isInteger, this.intValue, this.stringValue);

  static LinkNonce int_(int value) => LinkNonce._(true, value, null);

  static LinkNonce str(String value) {
    // ignore: unnecessary_null_comparison, dead_code
    if (value == null) throw ArgumentError.notNull('value');
    return LinkNonce._(false, 0, value);
  }

  @override
  bool operator ==(Object other) =>
      other is LinkNonce &&
      isInteger == other.isInteger &&
      intValue == other.intValue &&
      stringValue == other.stringValue;

  @override
  int get hashCode => Object.hash(isInteger, intValue, stringValue);

  @override
  String toString() => isInteger ? intValue.toString() : stringValue!;
}

/// The stable, never-reused link identity. Mirrors the GLP
/// `LinkId ::= link_id(Scheme, Endpoint, Nonce)`
/// (contracts/link-primitives.md §1, lines 51-56).
///
/// Value equality (record) is the idempotency-at-identity basis (FR-007): a
/// re-setup with the same [LinkId] must return the already-established
/// link rather than open a conflicting duplicate, so [LinkId] keys the
/// per-instance LinkId→handle registry built in T030. The components are all
/// ground (the GLP wrappers guard `ground/1` before the host sees them), so a
/// [LinkId] is always fully determined.
class LinkId {
  final LinkScheme scheme;
  final LinkAddress endpoint;
  final LinkNonce nonce;

  const LinkId(this.scheme, this.endpoint, this.nonce);

  @override
  bool operator ==(Object other) =>
      other is LinkId &&
      scheme == other.scheme &&
      endpoint == other.endpoint &&
      nonce == other.nonce;

  @override
  int get hashCode => Object.hash(scheme, endpoint, nonce);

  @override
  String toString() => 'link_id($scheme, $endpoint, $nonce)';
}

File: csharp/glp_link/seam/LinkId.cs
namespace GlpRuntime.Link.Seam;

/// <summary>
/// The per-establishment uniqueness nonce. Mirrors the GLP
/// <c>Nonce ::= Integer ; String</c> component
/// (contracts/link-primitives.md §1, line 56). The integer and string forms are
/// kept distinct (an integer <c>5</c> is NOT the string <c>"5"</c>) so the value
/// equality that backs idempotency-at-identity (FR-007) matches the GLP term
/// exactly rather than a lossy canonical text.
/// </summary>
public readonly record struct LinkNonce
{
    /// <summary>True when this is the <c>Integer</c> form; false for the <c>String</c> form.</summary>
    public bool IsInteger { get; }

    /// <summary>The integer value, valid only when <see cref="IsInteger"/>.</summary>
    public long IntValue { get; }

    /// <summary>The string value, valid only when <see cref="IsInteger"/> is false.</summary>
    public string? StringValue { get; }

    private LinkNonce(bool isInt, long i, string? s)
    {
        IsInteger = isInt;
        IntValue = i;
        StringValue = s;
    }

    public static LinkNonce Int(long value) => new(true, value, null);

    public static LinkNonce Str(string value)
    {
        if (value is null) throw new ArgumentNullException(nameof(value));
        return new LinkNonce(false, 0, value);
    }

    public override string ToString() => IsInteger ? IntValue.ToString() : StringValue!;
}

/// <summary>
/// The stable, never-reused link identity. Mirrors the GLP
/// <c>LinkId ::= link_id(Scheme, Endpoint, Nonce)</c>
/// (contracts/link-primitives.md §1, lines 51-56).
/// </summary>
/// <remarks>
/// Value equality (record) is the idempotency-at-identity basis (FR-007): a
/// re-setup with the same <see cref="LinkId"/> must return the already-established
/// link rather than open a conflicting duplicate, so <see cref="LinkId"/> keys the
/// per-instance LinkId→handle registry built in T030. The components are all
/// ground (the GLP wrappers guard <c>ground/1</c> before the host sees them), so a
/// <see cref="LinkId"/> is always fully determined.
/// </remarks>
public sealed record LinkId(LinkScheme Scheme, LinkAddress Endpoint, LinkNonce Nonce)
{
    public override string ToString() => $"link_id({Scheme}, {Endpoint}, {Nonce})";
}
